chasm/tree.go

```diff
@@ -153,6 +164,210 @@ func NewEmptyTree(
+				},
+				Attributes: &persistencespb.ChasmNodeMetadata_DataAttributes{
+					DataAttributes: &persistencespb.ChasmDataAttributes{},
+				},
+			},
+		}
+	case fieldTypeComponent:
+		n.serializedValue = &persistencespb.ChasmNode{
+			Metadata: &persistencespb.ChasmNodeMetadata{
+				InitialVersionedTransition: &persistencespb.VersionedTransition{
+					TransitionCount:          n.backend.NextTransitionCount(),
+					NamespaceFailoverVersion: n.backend.GetCurrentVersion(),
+				},
+				Attributes: &persistencespb.ChasmNodeMetadata_ComponentAttributes{
+					ComponentAttributes: &persistencespb.ChasmComponentAttributes{},
+				},
+			},
+		}
+	case fieldTypeComponentPointer:
+		panic("not implemented")
+	}
+	return nil
+}
+
+func validateValueType(v any) error {
+	return validateType(reflect.TypeOf(v))
+}
+
+func validateType(t reflect.Type) error {
+	// TODO: for component, interface should also be supported.
+	if t.Kind() != reflect.Ptr && t.Elem().Kind() != reflect.Struct {
+		return serviceerror.NewInternal("only pointer to struct is supported for tree node value")
+	}
+	return nil
+}
+
+func fieldName(f reflect.StructField) string {
+	if tagName := f.Tag.Get(fieldNameTag); tagName != "" {
+		return tagName
+	}
+	return f.Name
+}
+
+func (n *Node) setSerializedValue(
+	nodePath []string,
+	serializedNode *persistencespb.ChasmNode,
+) {
+	if len(nodePath) == 0 {
+		n.serializedValue = serializedNode
+		return
+	}
+
+	childName := nodePath[0]
+	childNode, ok := n.children[childName]
+	if !ok {
+		childNode = newNode(n.nodeBase, n, childName)
+		n.children[childName] = childNode
+	}
+	childNode.setSerializedValue(nodePath[1:], serializedNode)
+}
+
+// deserialize initializes the node's value from its serializedValue.
+// If value is of component type, it initializes every chasm.Field of it and sets node field but not value field
+// i.e. it doesn't deserialize recursively and must be called on every node separately.
+func (n *Node) deserialize(
+	valueT reflect.Type,
+) error {
+	if err := validateType(valueT); err != nil {
+		return err
+	}
+
+	if n.parent == nil {
+		// Top level node is always component.
+		return n.deserializeComponentNode(valueT)
+	}
+
+	switch n.serializedValue.GetMetadata().GetAttributes().(type) {
+	case *persistencespb.ChasmNodeMetadata_ComponentAttributes:
+		return n.deserializeComponentNode(valueT)
+	case *persistencespb.ChasmNodeMetadata_DataAttributes:
+		return n.deserializeDataNode(valueT)
+	case *persistencespb.ChasmNodeMetadata_CollectionAttributes:
+		panic("not implemented")
+	case *persistencespb.ChasmNodeMetadata_PointerAttributes:
+		// TODO: return serviceerror.NewInternal(...) instead.
+	}
+	return nil
+}
+
+func (n *Node) deserializeComponentNode(
+	valueT reflect.Type,
+) error {
+	// TODO: use n.serializedValue.GetComponentAttributes().GetType() instead to support deserialization to interface.
+	valueV := reflect.New(valueT.Elem())
+	if n.serializedValue == nil {
```

Comment:
When will this be nil? I think the metadata should always be available when this method is invoked.

Reply:
yes, after I moved to creation to `NewTree` it is always not `nil`. `Data` can be `nil` though.